File: backend/app/services/speedtest.py
"""WAN speed tests run from a container on the router itself.

Why a container
---------------
RouterOS has no internet speed test. ``/tool/speed-test`` and
``/tool/bandwidth-test`` both measure against *another RouterOS device*, which
answers a different question than "what is my line actually doing". Ookla's
``speedtest`` CLI does answer it, and RouterOS 7.4+ can run OCI containers, so
the test runs on the router - measuring the router's own WAN link, not the link
between the router and whatever machine MikroMan happens to be installed on.

How it is driven, given the REST API's limits
---------------------------------------------
There is no ``docker exec`` over REST: ``/container/shell`` exists but is an
interactive console command. The speedtest image is built to run once and exit,
which turns out to fit the REST surface exactly:

1. ``POST /container/start`` with the container's id - it runs and exits.
2. Its stdout reaches the RouterOS log, provided a ``container`` logging action
   exists (RouterOS does not create one by default, so :meth:`ensure_logging`
   adds it).
3. ``GET /log`` is polled for lines on the ``container`` topic newer than the
   marker taken before the start.

So the whole cycle is three ordinary REST calls and some patience. Nothing here
requires SSH or the console.

The parser is deliberately tolerant. Ookla's human-readable output has changed
format between CLI releases and the exact spacing differs between builds, so
every field is matched independently and a line that cannot be understood is
skipped rather than failing the run. A test that produced a download figure and
nothing else is still worth recording.
"""
import asyncio
import logging
import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

logger = logging.getLogger("mikroman.speedtest")

# The image the operator is offered. Purpose-built for RouterOS: 2.7 MiB,
# published for arm64 and amd64, runs once and exits.
DEFAULT_IMAGE = "quay.io/tangent/speedtest-cli:latest"
# Marks the container MikroMan manages, so an operator's own speedtest container
# is never started, stopped or removed by us.
CONTAINER_COMMENT = "mikroman:speedtest"
# RouterOS logging topic the container's stdout arrives on.
LOG_TOPIC = "container"
# A full Ookla run is ~15s of transfer plus server selection. Beyond this the
# run is reported as a timeout rather than left pending forever.
DEFAULT_TIMEOUT_SECONDS = 120.0
POLL_INTERVAL_SECONDS = 2.0


@dataclass
class SpeedTestReading:
    """What one run measured. Every field is optional - see the module docstring."""
    download_mbps: Optional[float] = None
    upload_mbps: Optional[float] = None
    ping_ms: Optional[float] = None
    jitter_ms: Optional[float] = None
    packet_loss_pct: Optional[float] = None
    server_name: Optional[str] = None
    isp: Optional[str] = None
    result_url: Optional[str] = None
    raw_output: str = ""
    status: str = "ok"
    error: Optional[str] = None
    lines: List[str] = field(default_factory=list)

    @property
    def has_any_figure(self) -> bool:
        return any(
            v is not None for v in
            (self.download_mbps, self.upload_mbps, self.ping_ms)
        )


# Ookla reports speeds in Mbps and may append a data-used note; latency lines
# carry a jitter figure in parentheses on newer builds. Each is matched on its
# own so a change to one line cannot break the others.
_PATTERNS = {
    "download_mbps": re.compile(r"Download:\s*([\d.]+)\s*Mbps", re.I),
    "upload_mbps": re.compile(r"Upload:\s*([\d.]+)\s*Mbps", re.I),
    "ping_ms": re.compile(r"(?:Idle )?Latency:\s*([\d.]+)\s*ms", re.I),
    "jitter_ms": re.compile(r"jitter:\s*([\d.]+)\s*ms", re.I),
    "packet_loss_pct": re.compile(r"Packet Loss:\s*([\d.]+)\s*%", re.I),
}
_SERVER = re.compile(r"Server:\s*(.+?)\s*$", re.I)
_ISP = re.compile(r"ISP:\s*(.+?)\s*$", re.I)
_RESULT_URL = re.compile(r"(https?://\S*speedtest\.net/result\S*)", re.I)
# Some builds emit "Result URL: ..." without the speedtest.net host.
_ANY_RESULT_URL = re.compile(r"Result\s*URL:\s*(\S+)", re.I)


def parse_speedtest_output(lines: List[str]) -> SpeedTestReading:
    """Pull figures out of Ookla CLI output, skipping anything unrecognised."""
    reading = SpeedTestReading(raw_output="\n".join(lines), lines=list(lines))

    for line in lines:
        for attr, pattern in _PATTERNS.items():
            if getattr(reading, attr) is None:
                match = pattern.search(line)
                if match:
                    try:
                        setattr(reading, attr, float(match.group(1)))
                    except ValueError:
                        pass
        if reading.server_name is None and "Server:" in line:
            match = _SERVER.search(line)
            if match:
                reading.server_name = match.group(1)[:200]
        if reading.isp is None and "ISP:" in line:
            match = _ISP.search(line)
            if match:
                reading.isp = match.group(1)[:200]
        if reading.result_url is None:
            match = _RESULT_URL.search(line) or _ANY_RESULT_URL.search(line)
            if match:
                reading.result_url = match.group(1)[:300]

    if not reading.has_any_figure:
        reading.status = "failed"
        reading.error = "Container produced no recognisable speed test output."
    return reading


class SpeedTestRunner:
    """Drives the speedtest container and reads its result back out of the log."""

    def __init__(self, client: Any):
        self.client = client

    # --- container lifecycle ------------------------------------------------

    async def find_container(self) -> Optional[Dict[str, Any]]:
        """The speedtest container MikroMan manages, if it has been created.

        Identified by our comment first. Falling back to the image name lets an
        operator who created the container by hand still use the button, which
        is friendlier than telling them their working container is the wrong one.
        """
        containers = await self.client.get_containers()
        for raw in containers:
            if (raw.get("comment") or "") == CONTAINER_COMMENT:
                return raw
        for raw in containers:
            if "speedtest" in (raw.get("remote-image") or raw.get("image") or "").lower():
                return raw
        return None

    async def ensure_logging(self) -> bool:
        """Make sure container output reaches the log we read results from.

        RouterOS ships no ``container`` logging action, so without this the
        container runs correctly and its output goes nowhere. Returns True when
        a rule is in place afterwards.
        """
        try:
            rules = await self.client.get_logging_rules()
        except Exception as e:
            logger.warning(f"Could not read logging rules: {e}")
            return False

        for rule in rules:
            topics = (rule.get("topics") or "").lower()
            if LOG_TOPIC in topics and (rule.get("disabled") or "false") != "true":
                return True

        try:
            await self.client.add_logging_rule(topics=LOG_TOPIC, action="memory")
            logger.info("Added a 'container' logging action so speed test output is readable")
            return True
        except Exception as e:
            logger.warning(f"Could not add container logging rule: {e}")
            return False

    async def create_container(
        self, *, interface: str, root_dir: str, image: str = DEFAULT_IMAGE
    ) -> Dict[str, Any]:
        """Create the speedtest container.

        ``start-on-boot`` is deliberately off: the image runs once and exits, so
        starting it at boot would produce one stray result per reboot and
        nothing else.
        """
        return await self.client.add_container({
            "remote-image": image,
            "interface": interface,
            "root-dir": root_dir,
            "comment": CONTAINER_COMMENT,
            "logging": "yes",
            "start-on-boot": "no",
        })

    # --- running ------------------------------------------------------------

    async def run(
        self,
        *,
        timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS,
        poll_interval: float = POLL_INTERVAL_SECONDS,
    ) -> SpeedTestReading:
        """Start the container, wait for it to finish, and parse what it logged."""
        container = await self.find_container()
        if container is None:
            return SpeedTestReading(
                status="failed",
                error="No speed test container on this router yet.",
            )

        await self.ensure_logging()

        # Everything already in the log is somebody else's; only lines after
        # this marker belong to the run we are about to start. Using the newest
        # existing id rather than a timestamp avoids depending on the router's
        # clock agreeing with ours.
        try:
            before = await self._container_log_ids()
        except Exception as e:
            logger.warning(f"Could not read the log before starting: {e}")
            before = set()

        try:
            await self.client.container_command("start", container[".id"])
        except Exception as e:
            return SpeedTestReading(status="failed", error=f"Could not start the container: {e}")

        return await self._await_result(before, timeout_seconds, poll_interval)

    async def _container_log_ids(self) -> set:
        entries = await self.client.get_log(topics=LOG_TOPIC)
        return {e.get(".id") for e in entries if e.get(".id")}

    async def _await_result(
        self, before_ids: set, timeout_seconds: float, poll_interval: float
    ) -> SpeedTestReading:
        """Poll the log until the run has produced a figure, or time runs out.

        Returns as soon as both a download and an upload figure are present
        rather than waiting for the container to be observed stopped: the
        container's state flips back to stopped the moment it exits, and racing
        that transition would sometimes miss it entirely.
        """
        deadline = asyncio.get_event_loop().time() + timeout_seconds
        newest: List[str] = []

        while asyncio.get_event_loop().time() < deadline:
            await asyncio.sleep(poll_interval)
            try:
                entries = await self.client.get_log(topics=LOG_TOPIC)
            except Exception as e:
                logger.debug(f"Log poll failed, will retry: {e}")
                continue

            newest = [
                e.get("message") or ""
                for e in entries
                if e.get(".id") not in before_ids
            ]
            if not newest:
                continue

            reading = parse_speedtest_output(newest)
            if reading.download_mbps is not None and reading.upload_mbps is not None:
                return reading

        # Out of time. Whatever was logged is still worth keeping and parsing -
        # a run that measured download but timed out on upload beats nothing.
        reading = parse_speedtest_output(newest)
        if reading.has_any_figure:
            reading.status = "ok"
            reading.error = "Timed out before the test finished; figures are partial."
        else:
            reading.status = "timeout"
            reading.error = (
                f"No speed test output within {timeout_seconds:.0f}s. Check that the "
                f"container has an internet-facing veth interface and that the "
                f"'container' logging topic is enabled."
            )
        return reading

File: backend-go/internal/services/speedtest.go
package services

import (
	"context"
	"fmt"
	"log/slog"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/masseselsev/mikroman/internal/db"
	"github.com/masseselsev/mikroman/internal/routeros"
)

const (
	DefaultSpeedTestImage  = "quay.io/tangent/speedtest-cli:latest"
	SpeedTestComment       = "mikroman:speedtest"
	SpeedTestLogTopic      = "container"
	DefaultSpeedTimeout    = 120 * time.Second
	SpeedTestPollInterval  = 2 * time.Second
)

var (
	reDownload   = regexp.MustCompile(`(?i)Download:\s*([\d.]+)\s*Mbps`)
	reUpload     = regexp.MustCompile(`(?i)Upload:\s*([\d.]+)\s*Mbps`)
	rePing       = regexp.MustCompile(`(?i)(?:Idle\s+)?Latency:\s*([\d.]+)\s*ms|Ping:\s*([\d.]+)\s*ms`)
	reJitter     = regexp.MustCompile(`(?i)jitter:\s*([\d.]+)\s*ms`)
	rePacketLoss = regexp.MustCompile(`(?i)Packet Loss:\s*([\d.]+)\s*%`)
	reServer     = regexp.MustCompile(`(?i)Server:\s*(.+?)\s*$`)
	reISP        = regexp.MustCompile(`(?i)ISP:\s*(.+?)\s*$`)
	reResultURL  = regexp.MustCompile(`(?i)(https?://\S*speedtest\.net/result\S*)|Result\s*URL:\s*(\S+)`)
)

type SpeedTestReading struct {
	DownloadMbps  *float64 `json:"download_mbps"`
	UploadMbps    *float64 `json:"upload_mbps"`
	PingMs        *float64 `json:"ping_ms"`
	JitterMs      *float64 `json:"jitter_ms"`
	PacketLossPct *float64 `json:"packet_loss_pct"`
	ServerName    *string  `json:"server_name"`
	ISP           *string  `json:"isp"`
	ResultURL     *string  `json:"result_url"`
	RawOutput     string   `json:"raw_output"`
	Status        string   `json:"status"` // "ok", "failed", "timeout"
	Error         *string  `json:"error"`
}

func (r *SpeedTestReading) HasAnyFigure() bool {
	return r.DownloadMbps != nil || r.UploadMbps != nil || r.PingMs != nil
}

// ParseSpeedTestOutput extracts figures from Ookla CLI stdout lines.
func ParseSpeedTestOutput(lines []string) SpeedTestReading {
	reading := SpeedTestReading{
		RawOutput: strings.Join(lines, "\n"),
		Status:    "ok",
	}

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		if reading.DownloadMbps == nil {
			if m := reDownload.FindStringSubmatch(line); len(m) > 1 {
				if v, err := strconv.ParseFloat(m[1], 64); err == nil {
					reading.DownloadMbps = &v
				}
			}
		}

		if reading.UploadMbps == nil {
			if m := reUpload.FindStringSubmatch(line); len(m) > 1 {
				if v, err := strconv.ParseFloat(m[1], 64); err == nil {
					reading.UploadMbps = &v
				}
			}
		}

		if reading.PingMs == nil {
			if m := rePing.FindStringSubmatch(line); len(m) > 1 {
				valStr := m[1]
				if valStr == "" && len(m) > 2 {
					valStr = m[2]
				}
				if v, err := strconv.ParseFloat(valStr, 64); err == nil {
					reading.PingMs = &v
				}
			}
		}

		if reading.JitterMs == nil {
			if m := reJitter.FindStringSubmatch(line); len(m) > 1 {
				if v, err := strconv.ParseFloat(m[1], 64); err == nil {
					reading.JitterMs = &v
				}
			}
		}

		if reading.PacketLossPct == nil {
			if m := rePacketLoss.FindStringSubmatch(line); len(m) > 1 {
				if v, err := strconv.ParseFloat(m[1], 64); err == nil {
					reading.PacketLossPct = &v
				}
			}
		}

		if reading.ServerName == nil {
			if m := reServer.FindStringSubmatch(line); len(m) > 1 {
				srv := strings.TrimSpace(m[1])
				reading.ServerName = &srv
			}
		}

		if reading.ISP == nil {
			if m := reISP.FindStringSubmatch(line); len(m) > 1 {
				isp := strings.TrimSpace(m[1])
				reading.ISP = &isp
			}
		}

		if reading.ResultURL == nil {
			if m := reResultURL.FindStringSubmatch(line); len(m) > 1 {
				urlStr := m[1]
				if urlStr == "" && len(m) > 2 {
					urlStr = m[2]
				}
				urlStr = strings.TrimSpace(urlStr)
				reading.ResultURL = &urlStr
			}
		}
	}

	if !reading.HasAnyFigure() {
		reading.Status = "failed"
		errMsg := "Container produced no recognisable speed test output."
		reading.Error = &errMsg
	}

	return reading
}

// SpeedTestRunner coordinates speed test execution on RouterOS via container.
type SpeedTestRunner struct {
	client *routeros.Client
}

func NewSpeedTestRunner(client *routeros.Client) *SpeedTestRunner {
	return &SpeedTestRunner{client: client}
}

// FindContainer identifies the speedtest container by comment or image name.
func (r *SpeedTestRunner) FindContainer(ctx context.Context) (*routeros.Container, error) {
	containers, err := r.client.GetContainers(ctx)
	if err != nil {
		return nil, err
	}

	for _, c := range containers {
		if c.Comment == SpeedTestComment {
			return &c, nil
		}
	}

	for _, c := range containers {
		tagLower := strings.ToLower(c.Tag)
		nameLower := strings.ToLower(c.Name)
		if strings.Contains(tagLower, "speedtest") || strings.Contains(nameLower, "speedtest") {
			return &c, nil
		}
	}

	return nil, nil
}

// EnsureLogging ensures that RouterOS routes container stdout to the system log.
func (r *SpeedTestRunner) EnsureLogging(ctx context.Context) error {
	type logRule struct {
		ID       string `json:".id"`
		Topics   string `json:"topics"`
		Action   string `json:"action"`
		Disabled string `json:"disabled"`
	}

	var rules []logRule
	if err := r.client.Get(ctx, "/system/logging", &rules); err == nil {
		for _, rule := range rules {
			if strings.Contains(strings.ToLower(rule.Topics), SpeedTestLogTopic) &&
				rule.Disabled != "true" && rule.Disabled != "yes" {
				return nil
			}
		}
	}

	payload := map[string]string{
		"topics": SpeedTestLogTopic,
		"action": "memory",
	}
	var res interface{}
	return r.client.Put(ctx, "/system/logging", payload, &res)
}

// CreateContainer provisions the speed test container on the router.
func (r *SpeedTestRunner) CreateContainer(ctx context.Context, iface, rootDir, image string) error {
	if image == "" {
		image = DefaultSpeedTestImage
	}

	payload := map[string]interface{}{
		"remote-image":  image,
		"interface":     iface,
		"root-dir":      rootDir,
		"comment":       SpeedTestComment,
		"logging":       "yes",
		"start-on-boot": "no",
	}

	return r.client.CreateContainer(ctx, payload)
}

// Run starts the container, polls the log for output, and returns the result.
func (r *SpeedTestRunner) Run(ctx context.Context, timeout time.Duration) (SpeedTestReading, error) {
	if timeout <= 0 {
		timeout = DefaultSpeedTimeout
	}

	container, err := r.FindContainer(ctx)
	if err != nil {
		errMsg := fmt.Sprintf("Failed to query containers: %v", err)
		return SpeedTestReading{Status: "failed", Error: &errMsg}, err
	}
	if container == nil {
		errMsg := "No speed test container configured on this router."
		return SpeedTestReading{Status: "failed", Error: &errMsg}, nil
	}

	_ = r.EnsureLogging(ctx)

	// Snapshot existing container log IDs so previous runs don't get read
	beforeIDs := make(map[string]bool)
	if logs, err := r.client.GetLogs(ctx, SpeedTestLogTopic, 200); err == nil {
		for _, l := range logs {
			if l.ID != "" {
				beforeIDs[l.ID] = true
			}
		}
	}

	// Start container
	if err := r.client.RunContainerAction(ctx, container.ID, "start"); err != nil {
		errMsg := fmt.Sprintf("Could not start speed test container: %v", err)
		return SpeedTestReading{Status: "failed", Error: &errMsg}, err
	}

	// Poll log until both download and upload are available or deadline expires
	deadline := time.Now().Add(timeout)
	ticker := time.NewTicker(SpeedTestPollInterval)
	defer ticker.Stop()

	var newest []string

	for {
		select {
		case <-ctx.Done():
			errMsg := "Speed test canceled"
			return SpeedTestReading{Status: "failed", Error: &errMsg}, ctx.Err()
		case now := <-ticker.C:
			if now.After(deadline) {
				reading := ParseSpeedTestOutput(newest)
				if reading.HasAnyFigure() {
					reading.Status = "ok"
					msg := "Timed out before test completed; measurements are partial."
					reading.Error = &msg
				} else {
					reading.Status = "timeout"
					msg := fmt.Sprintf("No speed test output received within %ds. Check that the container has WAN connectivity and container logging is active.", int(timeout.Seconds()))
					reading.Error = &msg
				}
				return reading, nil
			}

			logs, err := r.client.GetLogs(ctx, SpeedTestLogTopic, 100)
			if err != nil {
				slog.Debug("Failed to poll container logs for speedtest", "err", err)
				continue
			}

			var fresh []string
			for _, l := range logs {
				if l.ID != "" && !beforeIDs[l.ID] {
					fresh = append(fresh, l.Message)
				}
			}
			newest = fresh

			if len(newest) > 0 {
				reading := ParseSpeedTestOutput(newest)
				if reading.DownloadMbps != nil && reading.UploadMbps != nil {
					return reading, nil
				}
			}
		}
	}
}

// ConvertToDBModel converts a SpeedTestReading to a db.SpeedTestResult model.
func (r *SpeedTestReading) ToDBModel(routerID int) *db.SpeedTestResult {
	res := &db.SpeedTestResult{
		RouterID:      routerID,
		CreatedAt:     time.Now().UTC(),
		DownloadMbps:  r.DownloadMbps,
		UploadMbps:    r.UploadMbps,
		PingMs:        r.PingMs,
		JitterMs:      r.JitterMs,
		PacketLossPct: r.PacketLossPct,
		Status:        r.Status,
		RawOutput:     r.RawOutput,
	}

	if r.ServerName != nil {
		res.ServerName = db.NewNullString(*r.ServerName)
	}
	if r.ISP != nil {
		res.ISP = db.NewNullString(*r.ISP)
	}
	if r.ResultURL != nil {
		res.ResultURL = db.NewNullString(*r.ResultURL)
	}
	if r.Error != nil {
		res.Error = db.NewNullString(*r.Error)
	}

	return res
}
